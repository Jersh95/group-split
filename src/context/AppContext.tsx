import UserFirestoreClient from "@/client/UserFirestoreClient";
import { Session } from "next-auth";
import { useSession } from "next-auth/react";
import { createContext, useEffect, useRef, useState } from "react";

type AppContextType = {
  initialized: boolean;
  user: User;
  setUser: (user: User) => Promise<void>;
};

export const AppContext = createContext<AppContextType>();

export const AppContextProvider = ({
  children,
}: {
  children: React.ReactNode;
}) => {
  const userFirestoreClient = new UserFirestoreClient();

  const { data: session } = useSession();

  const sessionSetRef = useRef(false);

  const [user, setUser] = useState<User | null>();
  const [initialized, setInitialized] = useState<boolean>(false);

  async function initialize(session: Session): Promise<void> {
    try {
      const existingUser = await userFirestoreClient.getUserByEmail(
        session.user.email
      );

      if (existingUser) {
        setUser(existingUser);
      } else {
        const createdUser = await userFirestoreClient.createUser(session);
        setUser(createdUser);
      }
    } catch (e) {
      console.error("Error initializing.", e);
      setUser(null);
    }

    setInitialized(true);
  }

  useEffect(() => {
    if (!sessionSetRef.current && session) {
      sessionSetRef.current = true;
      initialize(session);
    }
  }, [session]);

  const contextValue = {
    initialized,
    user,
    setUser,
  };

  return (
    <AppContext.Provider value={contextValue}>{children}</AppContext.Provider>
  );
};
