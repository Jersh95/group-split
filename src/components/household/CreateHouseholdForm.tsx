import HouseholdFirestoreClient from "@/client/HouseholdFirestoreClient";
import { AppContext } from "@/context/AppContext";
import { useContext, useState } from "react";
import { useRef } from "react";
import { Button, FloatingLabel, Form } from "react-bootstrap";

type CreateHouseholdFormProps = {};

export const CreateHouseholdForm =
  ({}: CreateHouseholdFormProps): JSX.Element => {
    const householdFirestoreClient = new HouseholdFirestoreClient();

    const [validated, setValidated] = useState(false);

    const householdNameRef = useRef();

    const { user } = useContext(AppContext);

    const onSubmit = async (e: React.FormEvent) => {
      console.log("clicked submit");
      e.preventDefault();
      const form = e.currentTarget;
      // console.log("form", form);
      // console.log("target", e.currentTarget);
      // console.log("checking", form.checkValidity());
      setValidated(true);
      // console.log(
      //   "householdNameRef.current.value",
      //   householdNameRef.current.value
      // );
      if (form.checkValidity()) {
        try {
          console.log("trying");
          const household = await householdFirestoreClient.createHousehold(
            user,
            householdNameRef.current.value
          );
        } catch (e) {
          console.error("Error creating household.", e);
        }
      }
    };

    return (
      <Form
        id="createHouseholdForm"
        onSubmit={onSubmit}
        noValidate
        validated={validated}
      >
        <Form.Group>
          <FloatingLabel
            controlId="householdName"
            label="Household Name"
            className="mb-3"
          >
            <Form.Control
              ref={householdNameRef}
              required
              type="text"
              placeholder="Household Name"
            />
            <Form.Control.Feedback type="invalid">
              Enter a name for the household.
            </Form.Control.Feedback>
          </FloatingLabel>
        </Form.Group>
        <Button type="submit" id="createHouseholdButton">
          Submit
        </Button>
      </Form>
    );
  };
