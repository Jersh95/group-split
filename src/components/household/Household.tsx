import { useSession } from "next-auth/react";
import { useContext, useRef } from "react";
import { Container, Row, Col, Button } from "react-bootstrap";
import { CreateHouseholdForm } from "./CreateHouseholdForm";
import { AppContext } from "@/context/AppContext";

type HouseholdProps = {};

export const Household = ({}: HouseholdProps): JSX.Element => {
  const { user } = useContext(AppContext);

  return (
    <Container>
      {user?.household ? (
        <h1>{user?.household?.name}</h1>
      ) : (
        <CreateHouseholdForm />
      )}
    </Container>
  );
};
