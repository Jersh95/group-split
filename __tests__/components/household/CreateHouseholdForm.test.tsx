import { CreateHouseholdForm } from "@/components/household/CreateHouseholdForm";
import { AppContext } from "@/context/AppContext";
import { waitFor } from "@testing-library/react";
import { mount, shallow } from "enzyme";
import { Simulate, act } from "react-dom/test-utils";

const mockHouseholdFirestoreClient = {
  getHousehold: jest.fn().mockImplementation(() => {}),
  createHousehold: jest.fn().mockImplementation(() => {}),
};

jest.mock("../../../src/client/HouseholdFirestoreClient", () => {
  return jest.fn().mockImplementation(() => {
    return mockHouseholdFirestoreClient;
  });
});

describe("CreateHouseholdForm", () => {
  it("renders successfully", () => {
    const user: User = {
      userId: "abc123",
      name: "Homer Simpson",
      email: "homer.simpson@jershmail.com",
      image: "some.url",
      household: null,
    };

    const component = mount(
      <AppContext.Provider value={{ user: user }}>
        <CreateHouseholdForm />
      </AppContext.Provider>
    );
    expect(component).toMatchSnapshot();
  });

  // TODO: fix test
  // it("renders and submits", async () => {
  //   const user: User = {
  //     userId: "abc123",
  //     name: "Homer Simpson",
  //     email: "homer.simpson@jershmail.com",
  //     image: "some.url",
  //     household: null,
  //   };

  //   // mockHouseholdFirestoreClient.createHousehold.mockResolvedValue{{}};

  //   const component = mount(
  //     <AppContext.Provider value={{ user: user }}>
  //       <CreateHouseholdForm />
  //     </AppContext.Provider>
  //   );

  //   // console.log("component", component.debug());

  //   let input = component.find("#householdName").first();
  //   expect(input).toBeDefined();
  //   const form = component.find("#createHouseholdForm").first();
  //   expect(form).toBeDefined();

  //   act(() => {
  //     // input.simulate("change", { target: { value: "Test Name" } });
  //     input.simulate("change", {
  //       target: { name: "householdName", value: "Test Name" },
  //     });
  //   });

  //   // await act(async () => {
  //   //   await form.simulate("submit");
  //   // });

  //   await waitFor(async () => {
  //     expect(component.find("#householdName").first().value).toEqual(
  //       "Test Name"
  //     );
  //     // expect(input.value).toEqual("Test Name");
  //     // expect(mockHouseholdFirestoreClient.createHousehold).toHaveBeenCalled();
  //   });
  // });
});
