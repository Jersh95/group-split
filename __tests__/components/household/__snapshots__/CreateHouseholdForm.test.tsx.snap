// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateHouseholdForm renders successfully 1`] = `ReactWrapper {}`;
